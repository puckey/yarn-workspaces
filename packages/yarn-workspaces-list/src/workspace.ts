export type Name = string;
export type AbsolutePath = string;
export type RelativePath = string;

export interface Workspace {
  name: Name;
  path: AbsolutePath;
  location: RelativePath;
  workspaceDependencies: Name[];
  mismatchedWorkspaceDependencies: Name[];
  transitiveWorkspaceDependencies: Name[];
}
