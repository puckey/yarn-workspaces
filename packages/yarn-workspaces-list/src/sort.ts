import toposort from 'toposort';
import { Name, Workspace } from './workspace';

export function topologicallySort(workspaces: Workspace[]): Workspace[] {
  const byName: Record<Name, Workspace> = {};
  const graph: Array<[Name, Name]> = [];

  for (const workspace of workspaces) {
    byName[workspace.name] = workspace;

    for (const dependency of workspace.workspaceDependencies) {
      graph.push([dependency, workspace.name]);
    }
  }

  const sorted: Name[] = toposort(graph);
  const orphans = workspaces.filter((workspace) => !sorted.includes(workspace.name));
  const sortedWorkspaces = orphans.concat(sorted.map((name) => byName[name]));

  return sortedWorkspaces;
}

export type Phase = Workspace[];

export function phasedSort(list: Workspace[]): Phase[] {
  const sorted = topologicallySort(list);

  const phases: Phase[] = [[]];
  let currentPhaseIndex = 0;
  for (const workspace of sorted) {
    const currentPhase = phases[currentPhaseIndex];
    const hasDependency = currentPhase.some((phaseWorkspace) =>
      workspace.workspaceDependencies.includes(phaseWorkspace.name)
    );

    if (!hasDependency) {
      currentPhase.push(workspace);
    } else {
      phases.push([workspace]);
      currentPhaseIndex = phases.length - 1;
    }
  }

  return phases;
}
