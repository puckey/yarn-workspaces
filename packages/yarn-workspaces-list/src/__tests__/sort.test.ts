import { join, relative } from 'path';
import { listWorkspaces } from '../';
import { phasedSort, topologicallySort } from '../sort';
import { Workspace } from '../workspace';

jest.mock('child_process');

test('should sort workspaces topologically', async () => {
  const cwd = join(__dirname, '../../__fixtures__');
  const list = await listWorkspaces({ cwd });

  expect(topologicallySort(list).map(normalize(cwd))).toMatchSnapshot();
});

test('should sort workspaces in topological phases', async () => {
  const cwd = join(__dirname, '../../__fixtures__');
  const list = await listWorkspaces({ cwd });

  expect(phasedSort(list).map((phase) => phase.map(normalize(cwd)))).toMatchSnapshot();
});

function normalize(cwd: string) {
  return (workspace: Workspace): Workspace => ({
    ...workspace,
    path: `absolute(${relative(cwd, workspace.path)})`,
  });
}
