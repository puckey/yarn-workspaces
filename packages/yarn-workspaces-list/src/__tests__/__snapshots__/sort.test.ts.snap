// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should sort workspaces in topological phases 1`] = `
Array [
  Array [
    Object {
      "location": "packages/a",
      "mismatchedWorkspaceDependencies": Array [],
      "name": "a",
      "path": "absolute(packages/a)",
      "transitiveWorkspaceDependencies": Array [],
      "workspaceDependencies": Array [],
    },
  ],
  Array [
    Object {
      "location": ".",
      "mismatchedWorkspaceDependencies": Array [],
      "name": "workspaces",
      "path": "absolute()",
      "transitiveWorkspaceDependencies": Array [],
      "workspaceDependencies": Array [
        "a",
      ],
    },
    Object {
      "location": "packages/b",
      "mismatchedWorkspaceDependencies": Array [],
      "name": "b",
      "path": "absolute(packages/b)",
      "transitiveWorkspaceDependencies": Array [],
      "workspaceDependencies": Array [
        "a",
      ],
    },
    Object {
      "location": "packages/c",
      "mismatchedWorkspaceDependencies": Array [],
      "name": "c",
      "path": "absolute(packages/c)",
      "transitiveWorkspaceDependencies": Array [],
      "workspaceDependencies": Array [
        "a",
      ],
    },
  ],
  Array [
    Object {
      "location": "projects/d",
      "mismatchedWorkspaceDependencies": Array [],
      "name": "d",
      "path": "absolute(projects/d)",
      "transitiveWorkspaceDependencies": Array [
        "a",
      ],
      "workspaceDependencies": Array [
        "b",
        "c",
      ],
    },
  ],
]
`;

exports[`should sort workspaces topologically 1`] = `
Array [
  Object {
    "location": "packages/a",
    "mismatchedWorkspaceDependencies": Array [],
    "name": "a",
    "path": "absolute(packages/a)",
    "transitiveWorkspaceDependencies": Array [],
    "workspaceDependencies": Array [],
  },
  Object {
    "location": ".",
    "mismatchedWorkspaceDependencies": Array [],
    "name": "workspaces",
    "path": "absolute()",
    "transitiveWorkspaceDependencies": Array [],
    "workspaceDependencies": Array [
      "a",
    ],
  },
  Object {
    "location": "packages/b",
    "mismatchedWorkspaceDependencies": Array [],
    "name": "b",
    "path": "absolute(packages/b)",
    "transitiveWorkspaceDependencies": Array [],
    "workspaceDependencies": Array [
      "a",
    ],
  },
  Object {
    "location": "packages/c",
    "mismatchedWorkspaceDependencies": Array [],
    "name": "c",
    "path": "absolute(packages/c)",
    "transitiveWorkspaceDependencies": Array [],
    "workspaceDependencies": Array [
      "a",
    ],
  },
  Object {
    "location": "projects/d",
    "mismatchedWorkspaceDependencies": Array [],
    "name": "d",
    "path": "absolute(projects/d)",
    "transitiveWorkspaceDependencies": Array [
      "a",
    ],
    "workspaceDependencies": Array [
      "b",
      "c",
    ],
  },
]
`;
